### ISSUE TITLE
aws_autoscaling_lifecycle_hook failure - not waiting on aws_iam_role_policy when it exists

### ISSUE THREAD

**Issue by rbowlby** [hashicorp/terraform#3648]
aws_autoscaling_lifecycle_hook does appear to be appropriately depend on the initial creation of the aws_iam_role resource. However, that role in turn requires has an aws_iam_role_policy resource defined.

Terraform attempts to create the aws_autoscaling_lifecycle_hook resource ahead of the creation of the role's policy resource (aws_iam_role_policy)...causing failure of the lifecycle_hook test message:

tld; aws_autoscaling_lifecycle_hook should require BOTH of these have completed:
aws_iam_role
aws_iam_role_policy

```
Error applying plan:

1 error(s) occurred:

* aws_autoscaling_lifecycle_hook.awesomeservice_lifecycle_hook: Error putting lifecycle hook: ValidationError: Unable to publish test message to notification target arn:aws:sns:us-west-1:666:awesomeservice_lifecycle_dev using IAM role arn:aws:iam::666:role/awesomeservice_asg_role_dev. Please check your target and role configuration and try to put lifecycle hook again.
	status code: 400, request id: f3c590de-7c5a-11e5-8f4d-c7bb45098453
```

**Comment by catsby** [comment]
Hey @rbowlby – do you have a configuration file that demonstrates this? I'd like to verify and check the planned graph output. We've seen issues with IAM resources in the past, where we'll issue commands to create and get a successful response, but other resources that then try to reference it receive a 404. AWS has documented that IAM Roles can take a few seconds to propagate:

- http://docs.aws.amazon.com/fr_fr/AWSEC2/latest/WindowsGuide/iam-roles-for-amazon-ec2.html#launch-instance-with-role-console

I'd like to verify that Terraform's graph is correct, and it's a propagation issue, versus Terraform actually creating things out of order. 

Thanks!

**Comment by rbowlby** [comment]
I do remember the output showing the creation of the aws_iam_role_policy resource AFTER the aws_autoscaling_lifecycle_hook. That would appear to mean that this is an issue of Terraform not waiting on that policy.

I should mention that the aws_autoscaling_lifecycle_hook resource only has a direct dependency on the aws_iam_role resource. However, being that the aws_iam_role isn't in a working state till the aws_iam_role_policy is created that it would make sense if the aws_autoscaling_lifecycle_hook depended on both.

**Comment by rbowlby** [comment]
Here's the relevant excerpt. Pointing it to an ASG resource of your choosing should suffice to allow an apply.

```
/* ###########################################################
                     Auto Scaling LifeCycle
   ########################################################### */

resource "aws_sns_topic" "awesome_service_lifecycle" {
    name = "awesome_service_lifecycle_${var.environment}"
}

resource "aws_iam_role" "awesome_service_asg_role" {
    name = "awesome_service_asg_role_${var.environment}"
    assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Effect": "Allow",
      "Principal": {
        "Service": "autoscaling.amazonaws.com"
      },
      "Sid": ""
    }
  ]
}
EOF
}

resource "aws_iam_role_policy" "awesome_service_asg_role_policy" {
    name = "awesome_service_asg_role_policy_${var.environment}"
    role = "${aws_iam_role.awesome_service_asg_role.id}"
    policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "sns:Publish"
      ],
      "Resource": "${aws_sns_topic.awesome_service_lifecycle.arn}"
    }
  ]
}
EOF
}

resource "aws_autoscaling_lifecycle_hook" "awesome_service_lifecycle_hook" {
    name = "awesome_service_lifecycle_hook_${var.environment}"
    autoscaling_group_name = "${aws_autoscaling_group.awesome_service.name}"
    default_result = "CONTINUE"
    heartbeat_timeout = 2000
    lifecycle_transition = "autoscaling:EC2_INSTANCE_TERMINATING"
    notification_target_arn = "${aws_sns_topic.awesome_service_lifecycle.arn}"
    role_arn = "${aws_iam_role.awesome_service_asg_role.arn}"
}
```

**Comment by catsby** [comment]
Hey @rbowlby – 

Thanks for the config! It was helpful. I can't see any explicit dependency mentioned there, so I'm not sure why we would expect the hook to be created _after_ the `aws_iam_role_policy`... there's no dependency there. I've attached the generated dependency graph to demonstrate that there is no actual dependency 

![out](https://cloud.githubusercontent.com/assets/61721/10830190/53f663a0-7e4c-11e5-8262-997f49f3a72c.png)


Using your config I did get the error though. The issue then seems to be either the role, or the role policy, reporting that they are successfully created, but they aren't actually ready for use... I'm trying to dig in and see exactly what or where. Unfortunately this is not the first issue like this (something made but not ready) from IAM resources :confused:

**Comment by catsby** [comment]
This is an unfortunate race. If I apply this plan from start to finish, it works fine, because the `hook` waits on the ASG to complete:

```hcl
provider "aws" {
    region = "us-west-2"
}

resource "aws_launch_configuration" "foobar" {
  image_id = "ami-dfc39aef"
  instance_type = "t2.micro"

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_autoscaling_group" "bar" {
  availability_zones = ["us-west-2a"]
  name = "foobar3-terraform-test"
  max_size = 3
  min_size = 3
  health_check_grace_period = 300
  health_check_type = "ELB"
  desired_capacity = 3
  force_delete = true
  termination_policies = ["ClosestToNextInstanceHour", "OldestLaunchConfiguration", "OldestInstance",
  "NewestInstance" ]
  launch_configuration = "${aws_launch_configuration.foobar.name}"
}


/* ###########################################################
                     Auto Scaling LifeCycle
   ########################################################### */

resource "aws_sns_topic" "awesome_service_lifecycle" {
    name = "awesome_service_lifecycle_dev"
}

resource "aws_iam_role" "awesome_service_asg_role" {
    name = "awesome_service_asg_role_dev"
    assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Effect": "Allow",
      "Principal": {
        "Service": "autoscaling.amazonaws.com"
      },
      "Sid": ""
    }
  ]
}
EOF
}

resource "aws_iam_role_policy" "awesome_service_asg_role_policy" {
    name = "awesome_service_asg_role_policy_dev"
    role = "${aws_iam_role.awesome_service_asg_role.id}"
    policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "sns:Publish"
      ],
      "Resource": "${aws_sns_topic.awesome_service_lifecycle.arn}"
    }
  ]
}
EOF
}

resource "aws_autoscaling_lifecycle_hook" "awesome_service_lifecycle_hook" {
    name = "awesome_service_lifecycle_hook_dev"
    autoscaling_group_name = "${aws_autoscaling_group.bar.name}"
    default_result = "CONTINUE"
    heartbeat_timeout = 2000
    lifecycle_transition = "autoscaling:EC2_INSTANCE_TERMINATING"
    notification_target_arn = "${aws_sns_topic.awesome_service_lifecycle.arn}"
    role_arn = "${aws_iam_role.awesome_service_asg_role.arn}"
}
```

You can see the dependency in the plan:

![out](https://cloud.githubusercontent.com/assets/61721/10833270/aaedb3d6-7e5e-11e5-8128-e90a072cfff3.png)

That wasn't shown in my prior plan above, because I applied the ASG separately. #3694 will "fix" this via a bandaid retry

**Comment by catsby** [comment]
I merged #3694, so I'm going to close this. Thanks!
